dev/ruff/main.go

```diff
@@ -0,0 +1,198 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/dagger/dagger/dev/ruff/internal/dagger"
+)
+
+var (
+	pythonVersion     = "3.11"
+	pythonImageRepo   = "docker.io/library/python"
+	pythonImageTag    = fmt.Sprintf("%s-slim", pythonVersion)
+	pythonImageDigest = "sha256:fc39d2e68b554c3f0a5cb8a776280c0b3d73b4c04b83dbade835e2a171ca27ef"
+	pythonImage       = pythonImageRepo + ":" + pythonImageTag + "@" + pythonImageDigest
+)
+
+// Ruff is a fast Python linter implemented in Rust
+type Ruff struct{}
+
+// Lint a Python codebase
+func (ruff Ruff) Lint(
+	ctx context.Context,
+	// The Python source directory to lint
+	source *dagger.Directory,
+) *LintRun {
+	return &LintRun{
+		Source: source,
+	}
+}
+
+// The result of running the Ruff lint tool
+type LintRun struct {
+	// +private
+	Source *dagger.Directory
+}
+
+// Return a JSON report file for this run
+func (run LintRun) Report() *dagger.File {
+	cmd := []string{
+		"ruff", "check",
+		"--exit-zero",
+		"--output-format", "json",
```

Comment:
Ruff supports outputting as junit, same as popular test suites, including Pytest which we use.

Isn't it better to have a common util that can parse that format and reuse with multiple tools?

Reply:
I can add an argument, or additional function, to produce a junit report in a follow-up PR. For internal use the format doesn't matter since we wrap it in a GraphQL API anyway.

Reply:
I was thinking more in the lines of reducing the custom work on multiple linting modules (for different tools) that parse different report formats. Since `junit` is a common format, it can reduce the work to create the Dagger "linting report entries" (and even tests) from that output accross different tools in a common dependent module.

Reply:
To be clear, it's just a suggestion for another time since the work to read Ruff's report is already done here.

Reply:
Ah I see. Yeah that makes sense. Sadly not *all* our tools support junit (I forget which one doesn't, but I remember researching exactly that). We could still do it for the others, though!